File: clients/rn/__tests__/engine/room-engine.test.ts
import { RoomEngine } from '../../src/engine/room/RoomEngine';
import { InMemoryEventJournal } from '../../src/engine/persistence/InMemoryEventJournal';

describe('RoomEngine', () => {
  let journal: InMemoryEventJournal;
  let engine: RoomEngine;

  beforeEach(() => {
    journal = new InMemoryEventJournal();
    engine = new RoomEngine('test', journal);
  });

  afterEach(() => {
    engine.shutdown();
  });

  it('creates a room and sets state', async () => {
    const result = await engine.send({
      type: 'create_room', name: 'Test Room', description: 'A test room.', zone: 'test',
    });
    expect(result.type).toBe('ok');
    expect(engine.state.name).toBe('Test Room');
    expect(engine.state.description).toBe('A test room.');
    expect(journal.eventCount('test')).toBeGreaterThan(0);
  });

  it('enter and leave room', async () => {
    await engine.send({ type: 'create_room', name: 'Room', description: 'D', zone: 'z' });
    await engine.send({
      type: 'enter_room', entityId: 'p1', entityName: 'Alice', entityType: 'player', fromDirection: 'south',
    });
    expect(engine.state.entities['p1']).toBeDefined();

    await engine.send({ type: 'leave_room', entityId: 'p1', entityName: 'Alice', direction: 'north' });
    expect(engine.state.entities['p1']).toBeUndefined();
  });

  it('rejects taking non-existent object', async () => {
    await engine.send({ type: 'create_room', name: 'Room', description: 'D', zone: 'z' });
    const result = await engine.send({ type: 'take_object', entityId: 'p1', objectName: 'ghost' });
    expect(result.type).toBe('rejected');
    if (result.type === 'rejected') {
      expect(result.code).toBe('not_found');
    }
  });

  it('rejects taking untakeable object', async () => {
    await engine.send({
      type: 'create_room', name: 'Room', description: 'D', zone: 'z',
      objects: [{ id: 'o1', name: 'pedestal', description: 'A stone pedestal.', takeable: false }],
    });
    const result = await engine.send({ type: 'take_object', entityId: 'p1', objectName: 'pedestal' });
    expect(result.type).toBe('rejected');
    if (result.type === 'rejected') {
      expect(result.code).toBe('not_takeable');
    }
  });

  it('takes a takeable object', async () => {
    await engine.send({
      type: 'create_room', name: 'Room', description: 'D', zone: 'z',
      objects: [{ id: 'o1', name: 'sword', description: 'A blade.', takeable: true }],
    });
    const result = await engine.send({ type: 'take_object', entityId: 'p1', objectName: 'sword' });
    expect(result.type).toBe('ok');
    expect(Object.keys(engine.state.objects)).toHaveLength(0);
  });

  it('say in room generates said event', async () => {
    await engine.send({ type: 'create_room', name: 'Room', description: 'D', zone: 'z' });
    await engine.send({ type: 'say_in_room', entityId: 'p1', entityName: 'Alice', text: 'Hello' });
    const events = journal.allEvents('test');
    expect(events.some(e => e.type === 'said')).toBe(true);
  });

  it('events are persisted to journal', async () => {
    await engine.send({ type: 'create_room', name: 'Room', description: 'D', zone: 'z' });
    await engine.send({
      type: 'enter_room', entityId: 'p1', entityName: 'Alice', entityType: 'player', fromDirection: 'south',
    });
    await engine.send({ type: 'say_in_room', entityId: 'p1', entityName: 'Alice', text: 'Hello' });

    const replayed = journal.allEvents('test');
    expect(replayed[0].type).toBe('room_created');
    expect(replayed[1].type).toBe('entity_entered');
    expect(replayed[2].type).toBe('said');
  });

  it('emits events to listeners', async () => {
    const events: any[] = [];
    engine.onEvent(e => events.push(e));

    await engine.send({ type: 'create_room', name: 'Room', description: 'D', zone: 'z' });
    await engine.send({ type: 'say_in_room', entityId: 'p1', entityName: 'Alice', text: 'Hello' });

    expect(events.some(e => e.type === 'room_created')).toBe(true);
    expect(events.some(e => e.type === 'said')).toBe(true);
  });
});

File: clients/kmp/shared/src/commonTest/kotlin/org/wyrdsekai/app/engine/room/RoomEngineTest.kt
package org.wyrdsekai.app.engine.room

import kotlinx.coroutines.launch
import kotlinx.coroutines.test.runTest
import kotlinx.coroutines.yield
import org.wyrdsekai.app.engine.InMemoryEventJournal
import org.wyrdsekai.app.engine.event.WorldEvent
import org.wyrdsekai.app.protocol.Exit
import org.wyrdsekai.app.protocol.RoomObject
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue

class RoomEngineTest {

    @Test
    fun createRoomSetsState() = runTest {
        val journal = InMemoryEventJournal()
        val engine = RoomEngine("test", journal, null, null, this)

        val response = engine.send(RoomEngineCommand.CreateRoom(
            name = "Test Room",
            description = "A test room.",
            zone = "test",
        ))

        assertTrue(response is RoomEngineResponse.Ok)
        assertEquals("Test Room", engine.state.value.name)
        assertEquals("A test room.", engine.state.value.description)
        assertTrue(journal.eventCount("test") > 0)
        engine.shutdown()
    }

    @Test
    fun enterAndLeaveRoom() = runTest {
        val journal = InMemoryEventJournal()
        val engine = RoomEngine("test", journal, null, null, this)

        engine.send(RoomEngineCommand.CreateRoom("Room", "Desc", "zone"))
        engine.send(RoomEngineCommand.EnterRoom("p1", "Alice", "player", "north"))

        assertEquals(1, engine.state.value.entities.size)
        assertEquals("Alice", engine.state.value.entities["p1"]!!.name)

        engine.send(RoomEngineCommand.LeaveRoom("p1", "Alice", "south"))
        assertTrue(engine.state.value.entities.isEmpty())
        engine.shutdown()
    }

    @Test
    fun takeNonexistentObjectRejected() = runTest {
        val journal = InMemoryEventJournal()
        val engine = RoomEngine("test", journal, null, null, this)

        engine.send(RoomEngineCommand.CreateRoom("Room", "Desc", "zone"))
        val result = engine.send(RoomEngineCommand.TakeObject("p1", "ghost"))
        assertTrue(result is RoomEngineResponse.Rejected)
        assertEquals("not_found", (result as RoomEngineResponse.Rejected).code)
        engine.shutdown()
    }

    @Test
    fun takeUntakeableObjectRejected() = runTest {
        val journal = InMemoryEventJournal()
        val engine = RoomEngine("test", journal, null, null, this)

        engine.send(RoomEngineCommand.CreateRoom(
            name = "Room", description = "Desc", zone = "zone",
            objects = listOf(RoomObject("obj-1", "pedestal", "A stone pedestal", false)),
        ))
        val result = engine.send(RoomEngineCommand.TakeObject("p1", "pedestal"))
        assertTrue(result is RoomEngineResponse.Rejected)
        assertEquals("not_takeable", (result as RoomEngineResponse.Rejected).code)
        engine.shutdown()
    }

    @Test
    fun takeTakeableObjectSucceeds() = runTest {
        val journal = InMemoryEventJournal()
        val engine = RoomEngine("test", journal, null, null, this)

        engine.send(RoomEngineCommand.CreateRoom(
            name = "Room", description = "Desc", zone = "zone",
            objects = listOf(RoomObject("obj-1", "sword", "A sharp sword", true)),
        ))
        val result = engine.send(RoomEngineCommand.TakeObject("p1", "sword"))
        assertTrue(result is RoomEngineResponse.Ok)
        assertTrue(engine.state.value.objects.isEmpty())
        engine.shutdown()
    }

    @Test
    fun notificationsEmitted() = runTest {
        val journal = InMemoryEventJournal()
        val engine = RoomEngine("test", journal, null, null, this)

        // Verify events are persisted to journal (more reliable than SharedFlow timing)
        engine.send(RoomEngineCommand.CreateRoom("Room", "Desc", "zone"))
        engine.send(RoomEngineCommand.SayInRoom("p1", "Alice", "Hello"))

        val replayed = journal.allEvents("test")
        assertTrue(replayed.any { it is WorldEvent.RoomCreated })
        assertTrue(replayed.any { it is WorldEvent.Said })
        engine.shutdown()
    }

    @Test
    fun createRoomWithExitsAndObjects() = runTest {
        val journal = InMemoryEventJournal()
        val engine = RoomEngine("test", journal, null, null, this)

        engine.send(RoomEngineCommand.CreateRoom(
            name = "Nexus",
            description = "A hub.",
            zone = "foundation",
            exits = listOf(Exit("north", "terminal", "To Terminal")),
            objects = listOf(RoomObject("obj-1", "crystal", "A crystal", false)),
        ))

        assertEquals(1, engine.state.value.exits.size)
        assertEquals("terminal", engine.state.value.exits["north"]!!.targetRoom)
        assertEquals(1, engine.state.value.objects.size)
        assertEquals("crystal", engine.state.value.objects["obj-1"]!!.name)
        engine.shutdown()
    }

    @Test
    fun useNonexistentObjectRejected() = runTest {
        val journal = InMemoryEventJournal()
        val engine = RoomEngine("test", journal, null, null, this)

        engine.send(RoomEngineCommand.CreateRoom("Room", "Desc", "zone"))
        val result = engine.send(RoomEngineCommand.UseObject("p1", "ghost", null))
        assertTrue(result is RoomEngineResponse.Rejected)
        engine.shutdown()
    }

    @Test
    fun eventsPersisted() = runTest {
        val journal = InMemoryEventJournal()
        val engine = RoomEngine("test", journal, null, null, this)

        engine.send(RoomEngineCommand.CreateRoom("Room", "Desc", "zone"))
        engine.send(RoomEngineCommand.EnterRoom("p1", "Alice", "player", "north"))
        engine.send(RoomEngineCommand.SayInRoom("p1", "Alice", "Hello"))

        val replayed = journal.allEvents("test")
        assertTrue(replayed.size >= 3)
        assertTrue(replayed[0] is WorldEvent.RoomCreated)
        assertTrue(replayed[1] is WorldEvent.EntityEntered)
        assertTrue(replayed[2] is WorldEvent.Said)
        engine.shutdown()
    }
}
